master/src/test/java/org/apache/celeborn/service/deploy/master/SlotsAllocatorRackAwareSuiteJ.java

```diff
@@ -140,4 +148,310 @@ public void accept(WorkerInfo workerInfo) {
+    }
+
+    WorkersSupplier(String message, List<WorkerInfo> workerList) {
+      this.message = message;
+      this.workerList = Collections.unmodifiableList(workerList);
+    }
+
+    private static List<WorkerInfo> equalHostsPerRackWorkers(int numHostsPerRack, int numRacks) {
+      List<WorkerInfo> workers = new ArrayList<>(numRacks * numHostsPerRack);
+      for (int rack = 1; rack <= numRacks; rack++) {
+        final String rackName = "/rack/r" + rack;
+        for (int host = 1; host <= numHostsPerRack; host++) {
+          final String hostName = "h" + host + "r" + rack;
+          workers.add(initWorker(hostName, rackName));
+        }
+      }
+      return workers;
+    }
+
+    private static List<WorkerInfo> specifiedHostsPerRackWorkers(List<Integer> hostsPerRack) {
+      List<WorkerInfo> workers = new ArrayList<>();
+
+      int rack = 0;
+      for (int numHosts : hostsPerRack) {
+        rack++;
+        final String rackName = "/rack/r" + rack;
+        for (int host = 1; host <= numHosts; host++) {
+          final String hostName = "h" + host + "r" + rack;
+          workers.add(initWorker(hostName, rackName));
+        }
+      }
+
+      return workers;
+    }
+
+    @Override
+    public List<WorkerInfo> get() {
+      List<WorkerInfo> result = new ArrayList<>(workerList);
+      Collections.shuffle(result, RAND);
+      return result;
+    }
+
+    @Override
+    public String toString() {
+      return "WorkersSupplier{" + message + "}";
+    }
+
+    private static WorkerInfo initWorker(String host, String rack) {
+      long assumedPartitionSize = 64 * 1024 * 1024;
+      DiskInfo diskInfo = new DiskInfo("/mnt/a", 1024L * 1024L * 1024L * 1024L, 1, 1, 0);
+      diskInfo.maxSlots_$eq(diskInfo.actualUsableSpace() / assumedPartitionSize);
+
+      Map<String, DiskInfo> diskInfoMap = new HashMap<>(2);
+      diskInfoMap.put(diskInfo.mountPoint(), diskInfo);
+
+      WorkerInfo workerInfo =
+          new WorkerInfo(host, 1, 2, 3, 4, Collections.unmodifiableMap(diskInfoMap), null);
+      workerInfo.networkLocation_$eq(rack);
+      return workerInfo;
+    }
+  }
+
+  private static class SlotReplicaAllocatorTestCase {
+
+    private final Supplier<List<WorkerInfo>> workersSupplier;
+    // Total number of partitions
+    private final int numPartitions;
+
+    // The maximum number of slots per replica we are expecting
+    private final int expectedMaxSlotsPerHost;
+
+    SlotReplicaAllocatorTestCase(int numHostsPerRack, int numRacks, int numPartitions) {
+      this(
+          new WorkersSupplier(numHostsPerRack, numRacks),
+          numPartitions,
+          (int) Math.ceil(((double) numPartitions) / (numHostsPerRack * numRacks)));
+    }
+
+    SlotReplicaAllocatorTestCase(List<Integer> hostsPerRack, int numPartitions) {
+      this(
+          new WorkersSupplier(hostsPerRack),
+          numPartitions,
+          // 1 + (maxHostsPerRack * numPartitions) / (minHostsPerRack * totalHosts)
```

Comment:
IMO `expectedMaxSlotsPerHost` can be further narrowed to `1 + (numPartitions / totalHosts) * (1 + (maxHostsPerRack - secondMaxhostsPerRack + 1) / (totalHosts - maxHostsPerRack))`, which means we need to amortize `maxHostsPerRack - secondMaxhostsPerRack + 1` partitions to other hosts.

I tested with the following code and `testRackAwareRoundRobinReplicaDistribution` passes:
```
    SlotReplicaAllocatorTestCase(List<Integer> hostsPerRack, int numPartitions) {
      Collections.sort(hostsPerRack);
      int maxHostsPerRack = hostsPerRack.get(hostsPerRack.size() - 1);
      int secondMaxHostsPerRack = hostsPerRack.get(hostsPerRack.size() - 2);
      int totalHosts = hostsPerRack.stream().mapToInt(Integer::intValue).sum();
      int expected =
          (int)
              Math.ceil(
                  ((double) (numPartitions)
                      / totalHosts
                      * (1
                          + ((double) (maxHostsPerRack - secondMaxHostsPerRack + 1))
                              / (totalHosts - maxHostsPerRack))));
      this.workersSupplier = new WorkersSupplier(hostsPerRack);
      this.numPartitions = numPartitions;
      this.expectedMaxSlotsPerHost = expected;
    }
```

Reply:
Will update, thanks !